Examples/drawing-tests.lisp

```diff
@@ -1,100 +1,356 @@
 (in-package :clim-demo)
 
+(defparameter *drawing-tests-categories* nil)
 (defparameter *drawing-tests* (make-hash-table :test 'equal))
 
 (defparameter *width* 500)
 (defparameter *height* 700)
 (defparameter *border-width* 5)
 
-(defstruct drawing-test name description drawer)
+(defstruct drawing-test category name description display-function)
 
-(defmacro define-drawing-test (name arglist description &body body)
+(defun drawing-test-keyname (category name)
+  (concatenate 'string category "-" name))
+
+(defmacro define-drawing-test (category name (frame stream &rest arglist) description &body body)
+  (check-type category string)
   (check-type name string)
   (check-type description string)
-  `(setf (gethash ,name *drawing-tests*)
-         (make-drawing-test :name ,name
-                         :description ,description
-                         :drawer (lambda ,arglist ,@body))))
+  (alexandria:with-gensyms (g-category g-name)
+    `(let ((,g-name ,name)
+           (,g-category ,category))
+       (setf (gethash (drawing-test-keyname ,g-category ,g-name) *drawing-tests*)
+             (make-drawing-test :category ,g-category
+                                :name ,g-name
+                                :description ,description
+                                :display-function (lambda (,frame ,stream ,@arglist) ,@body)))
+       (pushnew ,g-category *drawing-tests-categories* :test #'string-equal))))
+
+(defun get-display (category name)
+  (let ((drawing-test (gethash (drawing-test-keyname category name) *drawing-tests*)))
+    (when drawing-test
+      (drawing-test-display-function drawing-test))))
 
 (define-application-frame drawing-tests ()
   ((recording-p :initform t)
    (signal-condition-p :initform nil)
-   (current-selection :initform nil))
+   (current-selection :initform nil)
+   (application-frame-backend :initform :clx-ttf)
+   (frames :initform (make-hash-table))
+   (application-frame-recording-p :initform nil))
```

Comment:
why do we have two different settings for recording?

Reply:
The second recording option is specific to the frame created with the "Use separate application frame" button.